### PR TITLE
Refactor(subtitle): subtitle rework

### PR DESCRIPTION
**Pull request by YvonTre**
# Refactor(subtitle): subtitle rework

- Refactor subtitle module with @Pat1enceLos .
- Modularize subtitle functionality into four subtitle modules for better scalability, readability and robustness.
- Follow button --> vuex --> actual effect pattern.

## `Subtitle.js`

- Every subtitle's base class.
- Manage subtitle's load and parse.
- A subclass of EventEmitter.

## `SubtitleManager`

- Manage subtitle's lifetime cycle and main functions.
- Manage premiumSubtitle's upload.
- Retrieve subtitle's meta info.

## `SubtitleLoader`

- Manage subtitle's cue renderers.
- Calculate the time some subtitle been played.

## `CueRenderer`

- Render cues at some time.

## `PremiumSubtitles`

- Break video into several segments based on current subtitle,
  - e.g. (subtitle)[[2.3, 4.5], [5.6, 9.8]...] --> (segments)[[0, 2.3], [2.3, 4.5], [4.5, 5.6], [5.6, 9.8]...].
- Every time currentTime update, check currentSegment change or not and then detemine a segment is finished or not.
- if a segment is finished, it won't be set as currentSegment any more during the subtitle's lifecycle.

## Current Problems/To-do

- [x] Vtt rendering default position is not right.
- [ ] ASS \pos tags is not properly supported.
- [ ] More ASS effects.
- [ ] Refactor Subtitle class as SrtSubtitle, AssSubtitle and so on.
- [ ] Save subtitle's segments for a proper period of time.

### CODE REVIEW

**Review by huzhenghui37 — DISMISSED** [review]

File: src/renderer/components/PlayingView/TheVideoController.vue [hunk 1]
```diff
@@ -444,11 +444,16 @@ export default {
...
       let isAdvanceColumeItem;
+      let isSubtitleScrollItem;
       const nodeList = document.querySelector('.advance-column-items').childNodes;
+      const subList = document.querySelector('.subtitle-scroll-items').childNodes;
       for (let i = 0; i < nodeList.length; i += 1) {
         isAdvanceColumeItem = nodeList[i].contains(event.target);
       }
-      if (!isAdvanceColumeItem) {
+      for (let i = 0; i < subList.length; i += 1) {
+        isSubtitleScrollItem = subList[i].contains(event.target);
```
Comment (ipy) [inline src/renderer/components/PlayingView/TheVideoController.vue:455]: 这里不需要 `break`?

File: src/renderer/components/Subtitle/Subtitle.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+import { parse as srtVttParser, toMS } from 'subtitle';
+import { compile as assParser } from 'ass-compiler';
+
+import Sagi from '@/helpers/sagi';
+
+const getSubtitleType = (src) => {
+  if (existsSync(src)) {
+    return 'local';
+  } else if (/^[a-f0-9]*/i.test(src)) { // need to know subtitle hash pattern
```
Comment (ipy) [inline src/renderer/components/Subtitle/Subtitle.js]: 这个能解释得详细一点吗？

Comment (YvonTre) [reply src/renderer/components/Subtitle/Subtitle.js]: 这个是判断src是本地文件还是在线字幕，在线字幕传进来的是字幕的hash，使用正则去匹配。

**Review by huzhenghui37 — DISMISSED** [review]

File: src/renderer/components/PlayingView/TheVideoController.vue (hunk 1 above)
Comment (ipy) [reply src/renderer/components/PlayingView/TheVideoController.vue:455]: `break`为啥不放后面...

File: src/renderer/components/Subtitle/Subtitle.js (hunk 2 above)
Comment (ipy) [reply src/renderer/components/Subtitle/Subtitle.js]: 那我理解下来是想匹配所有字符都是 a-f0-9 的字符串？这个正则其实能够任意字符串。

Comment (YvonTre) [reply src/renderer/components/Subtitle/Subtitle.js]: 正则写错了。。。应该是`/^[a-f0-9]+$/`的

File: src/renderer/components/PlayingView/TheVideoController.vue (hunk 1 above)
Comment (YvonTre) [reply src/renderer/components/PlayingView/TheVideoController.vue:455]: 现在把break放在后面了。

**Review by ipy — APPROVED** [review]

**Pull request merged**